File: backend/internal/ip/ip_test.go
package ip

import (
	"testing"
)

func TestExtractClientIP(t *testing.T) {
	tests := []struct {
		name    string
		headers map[string]string
		want    string
	}{
		{
			name:    "cloudfront IPv4 with port",
			headers: map[string]string{"cloudfront-viewer-address": "198.51.100.10:46532"},
			want:    "198.51.100.10",
		},
		{
			name:    "cloudfront unbracketed IPv6 with port",
			headers: map[string]string{"cloudfront-viewer-address": "2001:db8::1:46532"},
			want:    "2001:db8::1",
		},
		{
			name:    "x-forwarded-for multiple entries",
			headers: map[string]string{"x-forwarded-for": "203.0.113.7, 70.41.3.18, 150.172.238.178"},
			want:    "203.0.113.7",
		},
		{
			name:    "no relevant headers",
			headers: map[string]string{},
			want:    "unknown",
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := ExtractClientIP(tc.headers)
			if got != tc.want {
				t.Errorf("ExtractClientIP(%v) = %q; want %q", tc.headers, got, tc.want)
			}
		})
	}
}

File: backend/test/ingest/ip.test.ts
import { describe, it, expect } from 'vitest';
import { extractClientIp } from '../../src/ingest/ip';

describe('extractClientIp', () => {
  it('prefers CloudFront-Viewer-Address, stripping the port (IPv4)', () => {
    expect(extractClientIp({ 'cloudfront-viewer-address': '198.51.100.10:46532' }))
      .toBe('198.51.100.10');
  });
  it('handles IPv6 CloudFront-Viewer-Address (port after last colon)', () => {
    expect(extractClientIp({ 'cloudfront-viewer-address': '2001:db8::1:46532' }))
      .toBe('2001:db8::1');
  });
  it('falls back to the first X-Forwarded-For entry', () => {
    expect(extractClientIp({ 'x-forwarded-for': '203.0.113.7, 70.41.3.18, 150.172.238.178' }))
      .toBe('203.0.113.7');
  });
  it('returns "unknown" when no IP headers present', () => {
    expect(extractClientIp({})).toBe('unknown');
  });
});
